open-craft/opencraft
# AppServers Cleanup v2 - Only keep active, fallback and RC servers

Terminate app servers that were created more than `days` before now, except:
* the active appserver if there is one,
* a "release candidate" (rc) appserver, to allow testing before the next appserver activation (we keep the most recent running appserver)
* a fallback appserver, for `days` after activating an appserver, to allow reverts (we keep the most recent running appserver created before the current activation)

See OC-2041 . Follow-up to #143 

**Test Instructions**

1. Create an instance that is not associated with a PR:
   - Run `make shell`.
   - Create instance via:
     
     ``` python
     from instance.factories import instance_factory
     regular_instance = instance_factory(name="Regular instance, no PR", sub_domain="regular-instance-zohoRek7")
     ```
1. Spawn four app servers for the instance, faking their age:
     
     ``` python
     from datetime import timedelta
     from django.utils import timezone
     from freezegun import freeze_time

     with freeze_time(timezone.now() - timedelta(days=5)):
         oldest_appserver_id = regular_instance.spawn_appserver()

     with freeze_time(timezone.now() - timedelta(days=4)):
         fallback_appserver_id = regular_instance.spawn_appserver()

     with freeze_time(timezone.now() - timedelta(days=1)):
         active_appserver_id = regular_instance.spawn_appserver()

     with freeze_time(timezone.now() + timedelta(days=4)):
         rc_appserver_id = regular_instance.spawn_appserver()
     ```

1. Activate `active_appserver`:

     ``` python
     regular_instance.set_appserver_active(active_appserver_id)
     ```

1. Check the state of the appservers. You can use the GUI or do it directly in the shell via something like:
     
     ``` python
     for appserver in regular_instance.appserver_set.all():
         print(appserver)
         print("Status: ", appserver.status)
         print("Status (server): ", appserver.server.status)
     ```
     
     This should produce the following output:
     
     ```
     AppServer 1
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 2
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 3
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 4
     Status:  Running [running]
     Status (server):  Ready [ready]
     ```
1. Run the `clean_up` task:
     
     ``` python
     from instance.tasks import clean_up
     clean_up()
     ```
1. Verify that only the oldest appserver has been shut down. This should produce the following output:
     ```
     AppServer 1
     Status:  Terminated [terminated]
     Status (server):  Terminated [terminated]
     AppServer 2
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 3
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 4
     Status:  Running [running]
     Status (server):  Ready [ready]
     ```

1. Re-run the `cleanup_task` from different dates to check that the behavior correctly matches the description at the top of this PR. Especially:
     - The active appserver and the most recent running appserver (rc) should never be terminated
     - The fallback appserver should only be terminated after `days` have passed
     - Far in the future, only the active appserver and the most recent running appserver should remain

     To run the `cleanup_task` from different dates, use (make sure you use `HUEY_ALWAYS_EAGER=true`):
    
     ```
     with freeze_time(timezone.now() + timedelta(days=3)):
         clean_up()
     ```

     At D+3, you should get:
    
     ```
     AppServer 1
     Status:  Terminated [terminated]
     Status (server):  Terminated [terminated]
     AppServer 2
     Status:  Terminated [terminated]
     Status (server):  Terminated [terminated]
     AppServer 3
     Status:  Running [running]
     Status (server):  Ready [ready]
     AppServer 4
     Status:  Running [running]
     Status (server):  Ready [ready]
     ```

1. Ensure AppServers which aren't in the 'running' state always get deprovisioned after two days.

1. Double-check "OVH - OpenStack - Horizon - OpenCraft IM - Dev" account to see if correct set of app servers left.

**Reviewers**
